@here/harp-mapview/test/VisibleTileSetTest.ts

```diff
@@ -193,4 +204,31 @@ describe("VisibleTileSet", function() {
         assert(visibleTileDisposeSpies[1].notCalled);
         assert(parentDisposeSpy.calledOnce);
     });
+
+    it("caches frustum intersection for data sources with same tiling scheme", async function() {
+        setupBerlinCenterCameraFromSamples();
+        const zoomLevel = 15;
+        const storageLevel = 14;
+
+        const secondDataSource = new FakeOmvDataSource();
+        fixture.addDataSource(secondDataSource);
+
+        {
+            // Since the two data sources share same tiling scheme, frustum intersection will be
+            // calculated once, and resulting tiles will be the same for the two of them.
+            const result = updateRenderList(zoomLevel, storageLevel);
+            assert.equal(result.intersectionCount, 1);
+            assert.equal(result.tileList.length, 2);
+            expect(result.tileList[0].visibleTiles).to.eql(result.tileList[1].visibleTiles);
+        }
+
+        // If the second data source has a different tile scheme, the intersection must be
+        // calculated twice, once for each tiling scheme.
+        sinon.replace(secondDataSource, "getTilingScheme", sinon.stub().returns(hereTilingScheme));
+
+        {
+            const result = updateRenderList(zoomLevel, storageLevel);
+            assert.equal(result.intersectionCount, 2);
```

Comment:
I agree it's not a nice approach, but the main argument against it is that it exposes a useless piece of information in the return from VisibleTileSet. Overhead is probably not even measurable (we are talking about incrementing a counter).

The approach you suggest is too fragile, because it depends on implementation details (the number of calls to getSubTileKeys) to deduce what's been done (how many frustum intersections). It can be easily broken by implementation changes because the relation between there's no strong relation between those two pieces of information. If someone needs to call getSubTileKeys once more for any reason, they'll get a broken unit test and will have to figure out how to reimplement the test.

The only better way i can find of testing this is to actually refactor the whole frustum intersection calculation into a separate class and expose there the needed information. But I didn't think it was worth the effort.

Reply:
Fragility depends on what you're comparing to.
If you compare to hardcoded values, they i agree it may be very fragile.
If you compare to previous "runs", then it may be very stable.

Imagine scenario

```javascript
updateRenderList(..., [onlyOmvDataSource])
const originalCallsCount = getSubTileKeysStub.count();
getSubTileKeysStub.reset();
updateRenderList(..., [onlyOmvDataSource, backgrounDataSource])
assert.equals(getSubTileKeysStub.count(), originalCallsCount);
```
And next one:

```javascript
getSubTileKeysStub.reset();
updateRenderList(..., [onlyOmvDataSource, backgrounDataSource, someOtherDataSourceWithOtherTilingScheme])
assert.greaterThan(getSubTileKeysStub.count(), originalCallsCount);
```
This should give us pretty good insight if w reuse frustum calculations basing on the only observable behaviour - calls to `TilingScheme` interface.

I don't push though, i consider this as hint.

Reply:
Actually, i would count most expensive calls - my guess `getGeoBox & getWorldBox` whose perform lot of or math internally and probably have biggest impact on real performance.

Reply:
getSubTileKeys (as well as getGeoBox or getWorldBox) may be called anywhere else outside of getVisibleTilesForDataSources(). For instance, fillMissingTilesFromCache() is calling getSubTileKeys a non-obvious number of times and that gets executed in the unit test.
This forces to relax the assertions in the unit tests, as you show in your code snippet where you do
`assert.greaterThan(getSubTileKeysStub.count(), originalCallsCount);`

when you should be able to say:

`assert.equals(getSubTileKeysStub.count(), originalCallsCount+1);`

Reply:
Having also worked on this method, I wonder if it makes sense to split the VisibleTileSet into two classes, (or maybe one class and one free function), the computation of the TileKey's visible for the given DataSources doesn't need to be in the same class as the caching functionality etc. I think this discussion is showing that the current design is someone lacking.

Reply:
@nzjony @zbigg I think it makes sense to move computeFrustumIntersection() method. I can move it the the FrustumIntersection interface and make a full blown class. 
That way we can test directly frustum intersection code without having to go through the VIsibleTileSet class. Is that ok?